File: library-extensions/src/main/java/com/mikepenz/fastadapter_extensions/items/ThreeLineItem.java
package com.mikepenz.fastadapter_extensions.items;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.support.annotation.DrawableRes;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.mikepenz.fastadapter.items.AbstractItem;
import com.mikepenz.fastadapter.utils.ViewHolderFactory;
import com.mikepenz.library_extensions.R;
import com.mikepenz.materialize.holder.ImageHolder;
import com.mikepenz.materialize.holder.StringHolder;

/**
 * Created by fabianterhorst on 30.03.16.
 */
public class ThreeLineItem extends AbstractItem<ThreeLineItem, ThreeLineItem.ViewHolder> {

    private static final ViewHolderFactory<? extends ViewHolder> FACTORY = new ItemFactory();

    private StringHolder mName, mDescription;

    private ImageHolder mAvatar, mIcon;

    public ThreeLineItem withName(String name) {
        this.mName = new StringHolder(name);
        return this;
    }

    public ThreeLineItem withDescription(String description) {
        this.mDescription = new StringHolder(description);
        return this;
    }

    public ThreeLineItem withAvatar(Drawable avatar) {
        this.mAvatar = new ImageHolder(avatar);
        return this;
    }

    public ThreeLineItem withAvatar(@DrawableRes int avatarRes) {
        this.mAvatar = new ImageHolder(avatarRes);
        return this;
    }

    public ThreeLineItem withAvatar(Uri uri) {
        this.mAvatar = new ImageHolder(uri);
        return this;
    }

    public ThreeLineItem withAvatar(Bitmap bitmap) {
        this.mAvatar = new ImageHolder(bitmap);
        return this;
    }

    public ThreeLineItem withAvatar(String url) {
        this.mAvatar = new ImageHolder(Uri.parse(url));
        return this;
    }

    public ThreeLineItem withIcon(Drawable icon) {
        this.mIcon = new ImageHolder(icon);
        return this;
    }

    public ThreeLineItem withIcon(@DrawableRes int iconRes) {
        this.mIcon = new ImageHolder(iconRes);
        return this;
    }

    public ThreeLineItem withIcon(Uri uri) {
        this.mIcon = new ImageHolder(uri);
        return this;
    }

    public ThreeLineItem withIcon(Bitmap bitmap) {
        this.mIcon = new ImageHolder(bitmap);
        return this;
    }

    public StringHolder getName() {
        return mName;
    }

    public ImageHolder getAvatar() {
        return mAvatar;
    }

    public ImageHolder getIcon() {
        return mIcon;
    }

    @Override
    public int getType() {
        return R.id.three_line_item_id;
    }

    @Override
    public int getLayoutRes() {
        return R.layout.three_line_item;
    }


    @Override
    public void bindView(ViewHolder holder) {
        super.bindView(holder);
        mName.applyTo(holder.name);
        mDescription.applyTo(holder.description);
        if (mAvatar != null) {
            if (holder.avatar.getVisibility() == View.GONE) {
                holder.avatar.setVisibility(View.VISIBLE);
            }
            mAvatar.applyTo(holder.avatar);
        } else if (holder.avatar.getVisibility() == View.VISIBLE) {
            holder.avatar.setVisibility(View.GONE);
        }
        if (mIcon != null) {
            if (holder.icon.getVisibility() == View.GONE) {
                holder.icon.setVisibility(View.VISIBLE);
            }
            mIcon.applyTo(holder.icon);
        } else if (holder.icon.getVisibility() == View.VISIBLE) {
            holder.icon.setVisibility(View.GONE);
        }
    }

    protected static class ItemFactory implements ViewHolderFactory<ViewHolder> {
        public ViewHolder create(View v) {
            return new ViewHolder(v);
        }
    }

    @Override
    public ViewHolderFactory<? extends ViewHolder> getFactory() {
        return FACTORY;
    }

    protected static class ViewHolder extends RecyclerView.ViewHolder {

        protected TextView name, description;
        protected ImageView avatar, icon;

        public ViewHolder(View view) {
            super(view);
            name = (TextView) view.findViewById(R.id.name);
            description = (TextView) view.findViewById(R.id.description);
            avatar = (ImageView) view.findViewById(R.id.avatar);
            icon = (ImageView) view.findViewById(R.id.icon);
        }
    }
}

File: library-extensions-ui/src/main/java/com/mikepenz/fastadapter/ui/items/ThreeLineItem.kt
package com.mikepenz.fastadapter.ui.items

import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.net.Uri
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.DrawableRes
import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.fastadapter.items.AbstractItem
import com.mikepenz.fastadapter.ui.R
import com.mikepenz.fastadapter.ui.utils.FastAdapterUIUtils
import com.mikepenz.fastadapter.ui.utils.ImageHolder
import com.mikepenz.fastadapter.ui.utils.StringHolder

/**
 * Created by fabianterhorst on 30.03.16.
 */
class ThreeLineItem : AbstractItem<ThreeLineItem.ViewHolder>() {

    var name: StringHolder? = null
        private set

    var description: StringHolder? = null
        private set

    var avatar: ImageHolder? = null
        private set

    var icon: ImageHolder? = null
        private set

    override val type: Int
        get() = R.id.three_line_item_id

    override val layoutRes: Int
        get() = R.layout.three_line_item

    fun withName(name: String): ThreeLineItem {
        this.name = StringHolder(name)
        return this
    }

    fun withDescription(description: String): ThreeLineItem {
        this.description = StringHolder(description)
        return this
    }

    fun withAvatar(avatar: Drawable): ThreeLineItem {
        this.avatar = ImageHolder(avatar)
        return this
    }

    fun withAvatar(@DrawableRes avatarRes: Int): ThreeLineItem {
        this.avatar = ImageHolder(avatarRes)
        return this
    }

    fun withAvatar(uri: Uri): ThreeLineItem {
        this.avatar = ImageHolder(uri)
        return this
    }

    fun withAvatar(bitmap: Bitmap): ThreeLineItem {
        this.avatar = ImageHolder(bitmap)
        return this
    }

    fun withAvatar(url: String): ThreeLineItem {
        this.avatar = ImageHolder(Uri.parse(url))
        return this
    }

    fun withIcon(icon: Drawable): ThreeLineItem {
        this.icon = ImageHolder(icon)
        return this
    }

    fun withIcon(@DrawableRes iconRes: Int): ThreeLineItem {
        this.icon = ImageHolder(iconRes)
        return this
    }

    fun withIcon(uri: Uri): ThreeLineItem {
        this.icon = ImageHolder(uri)
        return this
    }

    fun withIcon(bitmap: Bitmap): ThreeLineItem {
        this.icon = ImageHolder(bitmap)
        return this
    }

    override fun bindView(holder: ViewHolder, payloads: List<Any>) {
        super.bindView(holder, payloads)
        if (isEnabled) {
            holder.itemView.setBackgroundResource(FastAdapterUIUtils.getSelectableBackground(holder.itemView.context))
        }
        name?.applyTo(holder.name)
        description?.applyTo(holder.description)
        ImageHolder.applyToOrSetGone(avatar, holder.avatar)
        ImageHolder.applyToOrSetGone(icon, holder.icon)
    }

    override fun unbindView(holder: ViewHolder) {
        holder.name.text = null
        holder.description.text = null
        holder.avatar.setImageDrawable(null)
        holder.avatar.visibility = View.VISIBLE
        holder.icon.setImageDrawable(null)
        holder.icon.visibility = View.VISIBLE
    }

    override fun getViewHolder(v: View): ViewHolder = ViewHolder(v)

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var name: TextView = view.findViewById<View>(R.id.name) as TextView
        var description: TextView = view.findViewById<View>(R.id.description) as TextView
        var avatar: ImageView = view.findViewById<View>(R.id.avatar) as ImageView
        var icon: ImageView = view.findViewById<View>(R.id.icon) as ImageView
    }
}
